docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,217 @@
+accept, reject, or completely ignore some or all of these transactions. This is
+an important consideration to make as the application can essentially define and
+control its own mempool allowing it to define sophisticated transaction priority
+and filtering mechanisms, by completely ignoring the `TxRecords` Tendermint
+sends it, favoring its own transactions. This essentially means that the Tendermint
+mempool acts more like a gossip data structure.
+
+The second ABCI method, `ProcessProposal`, is used to process the block proposer's
+proposal as defined by `PrepareProposal`. This ABCI method requests that the
+application evaluate the entire proposed block for validity.
+
+It is important to note that in this phase of ABCI++ integration, the application
+is NOT responsible for locking semantics -- Tendermint will still be responsible
+for that. In the future, however, the application will be responsible for locking,
+which allows for parallel execution possibilities.
+
+## Decision
+
+We will integrate the first phase of ABCI++, which will be introduced in Tendermint
+v0.37.0, in the next major release of the Cosmos SDK. We will integrate the two
+aforementioned ABCI++ methods on the `BaseApp` type. We describe the implementations
+of the two methods individually below.
+
+Prior to describing the implementation of the two new methods, it is important to
+note that the existing ABCI methods, `CheckTx`, `DeliverTx`, etc, still exist and
+serve the same functions as they do now.
+
+### `PrepareProposal`
+
+Prior to evaluating the decision for how to implement `PrepareProposal`, it is
+important to note that `CheckTx` will still be executed and will be responsible
+for evaluating transaction validity as it does now, with one very important
+_additive_ distinction.
+
+When executing transactions in `CheckTx`, the application will now add valid
+transactions, i.e. passing the AnteHandler, to it's own mempool data structure.
+In order to provide a flexible approach to meet the varying needs of application
+developers, we will define both a mempool interface and a data structure utilizing
+Golang generics, allowing developers to focus only on transaction
+ordering. Developers requiring absolute full control can implement their own
+custom mempool implementation.
+
+We define the general mempool interface as follows (subject to change):
+
+```go
+// MempoolTx we define a mempool transaction interface that is as minimal as
+// possible, only requiring applications to define the size of the transaction
+// to be used when reaping and getting the transaction itself. Interface type
+// casting can be used in the actual mempool implementation.
+type MempoolTx interface {
+	// Size returns the size of the transaction in bytes.
+	Size(codec.Codec) int
+	Tx() sdk.Tx
+}
+
+type Mempool[T MempoolTx] interface {
+	// Insert attempts to insert a MempoolTx into the mempool returning an error
+	// upon failure.
+	Insert(T) error
```

Comment:
Is this expected to place the transaction into the Tendermint mempool for gossiping?

Reply:
this would be the app side mempool. Tendermint gets the tx added for gossiping with a successful checktx response

Reply:
Right, my question is if the `Insert` call will also place the transaction into the Tendermint mempool or will this method only place the transaction into the app-side mempool and app developers need to separately call Tendermint to hand the transaction to Tendermint for gossiping.

In either case, the app-side mempool will need to handle getting `CheckTx` called for a transaction that's already in the app-side mempool, either because the app-side mempool called Tendermint or because the app developer did.

Reply:
the logic would be separate, don't think we should merge tendermint txpool and app side mempool logic. Checktx returning success will continue to handle tendermint insertion. The logic around checktx and delivertx doesn't change in regards to tendermint.

Reply:
Let me clarify, txs are added to the app-side mempool to during `CheckTx`, just the same as when they're added to Tendermint's mempool.

Reply:
@alexanderbez I'm assuming this API is developer facing and not just aimed at gluing Tendermint at the SDK mempool together. By developer facing, I mean that someone implementing an SDK application can add a transaction of their own creation to the app-side mempool. If that's the case, then there's some ambiguity as to whether or not this will cause the transaction to be gossiped or what the application developer should expect in terms of gossip.

If this API is developer facing, should developers that _just_ call `Insert` expect the transaction to be gossiped to other nodes?

Reply:
@williambanfield My understanding is that this API is developer facing, but the API is the interface. The developer is expected to implement this interface in order to configure their own custom app-side mempool.

My understanding is that `Insert` is not a method that is called by a Cosmos developer. Instead it's a method that the developer must _implement_ in order to receive transactions from Tendermint. The Cosmos SDK internals are the caller of `Insert`, not the developer. By implementing `Insert`, the developer can decide which data structures to use to store this transaction for later use in `ReapMaxBytes`.

Reply:
Thanks for clarifying @prettymuchbryce ! My understanding was a bit flipped then. If that's the case, then the primary way for an application to add a transaction to _this_ mempool is by adding it to the Tendermint mempool. This will trigger a `CheckTx` call which will subsequently call the `Insert` method.

Reply:
Yes, @prettymuchbryce is correct. However, there is nothing stopping the app from inserting custom txs when `Insert` is called during `PrepareProposal`.

Nothing is set in stone here. Looking for feedback. Perhaps we should offer more flexibility, i.e. some sort of "hook", in `PrepareProposal`?